### PR TITLE
[Perf][foundry][Pool] Fold the 2d and 3d pool windows into one output thread

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

- Rewrite the `MaxPool2d`, `MaxPool3d` and `AvgPool3d` forward kernels, values and indices alike, as one flat pass over output positions: a thread owns one output, folds its window into a register, and decodes its position with one division per spatial axis instead of one per axis plus two.
- Decide when the kernel is built whether a window can leave the input, so the shapes that cannot lose their bounds arithmetic entirely.
- Give a 2d value thread a `tile_h` by `tile_w` block of outputs, read the input band those outputs share once, and separate the window: row maxima across the band, then the maximum down them.
- Read that band as 16-byte vectors where the row's alignment allows it, which is the even-window shapes.
- Take the with-indices maximum the way PyTorch's own kernel takes it, with one predicate and a seeded position, instead of five accumulators that reproduce the same three properties the long way.
- Let `AvgPool3dSpatialKernel` delegate to the general kernel rather than carry a second copy of the scan.

## TileFoundry Description

```python
@module(
    entry="max_pool2d",
    target=CudaTarget("nvidia.h200_sxm"),
    topologies=(Topology("cta", 132),),
)
class ResnetStem:
    @func
    def max_pool2d(x: Tensor[(32, 64, 112, 112), "f16"]) -> Tensor[(32, 64, 56, 56), "f16"]:
        edge_h = tf.full_like(x[:, :, 0:1, :], value=NEG_INF)
        padded_h = tf.concat([edge_h, x, edge_h], axis=2)
        edge_w = tf.full_like(padded_h[:, :, :, 0:1], value=NEG_INF)
        padded = tf.concat([edge_w, padded_h, edge_w], axis=3)
        taps = tf.stack(
            [
                padded[:, :, t // 3 : t // 3 + 111 : 2, t % 3 : t % 3 + 111 : 2]
                for t in range(9)
            ],
            axis=-1,
        )
        return tf.reduce(taps, axes=(-1,), keepdim=False, kind="max")
```

The 3d modules are the same shape one axis up, with `kind="mean"` for the average pool.
`tilefoundry check` holds each delivered kernel to its authored program, run by the
evaluator: `--fn allclose --atol 1e-3 --rtol 1e-3` reports `max_violation 0` for
`f16[32,64,56,56]` (`max_pool2d`, `resnet-stem`), `f16[4,128,8,28,28]` (`max_pool3d`,
`c3d-pool2`) and `f16[1,32,8,28,28]` (`avg_pool3d`, `video-2x2x2`).

## Performance

| Environment | Value |
| --- | --- |
| image | ghcr.io/tile-ai/tileops-runner:cu132-torch2.13-tl-afcebed1-foundry-dev |
| digest | sha256:fe87c536154ad2ce3bffa4ff97e28fbfde7fcf6598d9d87c3f3dcc4ee9e44b5c |
| gpu | NVIDIA H200 |
| driver | 595.71.05 |
| cuda | 13.2 |
| torch | 2.13.0+cu132 |
| tilelang | 0.1.11+cu132.gitafcebed1 |
| tilefoundry | 0.1.dev131+g2306bf050 |
| timer | native CUPTI |

Method: candidate, incumbent and both torch baselines ran in one process on one idle
device, on common inputs carrying injected NaN and -inf. Each TileLang implementation was
searched over its own config space and is reported at its best; each was checked against
its `torch.nn.functional` reference — values and positions — before timing, then timed by
`bench_kernel` through CUPTI, which clears L2 before every iteration. `torch._dynamo.reset()`
precedes each `torch.compile`, without which later cases reuse an earlier compilation and
torch-compile is mistimed. Compilation and tuning were excluded; CUPTI failure was
fail-closed.

Ratio in comparator columns: implementation / candidate. &#x1F7E2; > 1 means the candidate
is faster; &#x1F534; <= 1 means it is not.

### MaxPool2dFwdOp

| Workload | Candidate (ms) | TileOPs incumbent (ms)<br>/ candidate | torch-compile (ms)<br>/ candidate | torch-ref (ms)<br>/ candidate |
| --- | ---: | ---: | ---: | ---: |
| resnet-stem float16 | 0.026816 | **0.037057<br>&#x1F7E2;&nbsp;1.3819x** | **0.030560<br>&#x1F7E2;&nbsp;1.1396x** | **0.106305<br>&#x1F7E2;&nbsp;3.9642x** |
| resnet-stem bfloat16 | 0.026624 | **0.037120<br>&#x1F7E2;&nbsp;1.3942x** | **0.030720<br>&#x1F7E2;&nbsp;1.1538x** | **0.106528<br>&#x1F7E2;&nbsp;4.0012x** |
| resnet-stem float32 | 0.035680 | **0.041920<br>&#x1F7E2;&nbsp;1.1749x** | **0.041472<br>&#x1F7E2;&nbsp;1.1623x** | **0.103873<br>&#x1F7E2;&nbsp;2.9112x** |
| vgg-block float16 | 0.006112 | **0.006768<br>&#x1F7E2;&nbsp;1.1073x** | **0.007840<br>&#x1F7E2;&nbsp;1.2827x** | **0.021696<br>&#x1F7E2;&nbsp;3.5497x** |
| vgg-block bfloat16 | 0.006112 | **0.006896<br>&#x1F7E2;&nbsp;1.1283x** | **0.006528<br>&#x1F7E2;&nbsp;1.0681x** | **0.021729<br>&#x1F7E2;&nbsp;3.5551x** |
| vgg-block float32 | 0.010368 | **0.010641<br>&#x1F7E2;&nbsp;1.0263x** | 0.009920<br>&#x1F534;&nbsp;0.9568x | **0.022017<br>&#x1F7E2;&nbsp;2.1236x** |
| alexnet-ceil float16 | 0.007744 | **0.008032<br>&#x1F7E2;&nbsp;1.0372x** | **0.008944<br>&#x1F7E2;&nbsp;1.1550x** | **0.026848<br>&#x1F7E2;&nbsp;3.4669x** |
| alexnet-ceil bfloat16 | 0.007680 | **0.007936<br>&#x1F7E2;&nbsp;1.0333x** | **0.009025<br>&#x1F7E2;&nbsp;1.1751x** | **0.026880<br>&#x1F7E2;&nbsp;3.5000x** |
| alexnet-ceil float32 | 0.011520 | **0.011872<br>&#x1F7E2;&nbsp;1.0306x** | **0.012992<br>&#x1F7E2;&nbsp;1.1278x** | **0.026784<br>&#x1F7E2;&nbsp;2.3250x** |

### MaxPool2dIndicesFwdOp

| Workload | Candidate (ms) | TileOPs incumbent (ms)<br>/ candidate | torch-compile (ms)<br>/ candidate | torch-ref (ms)<br>/ candidate |
| --- | ---: | ---: | ---: | ---: |
| resnet-stem float16 | 0.047969 | **0.080465<br>&#x1F7E2;&nbsp;1.6774x** | **0.053056<br>&#x1F7E2;&nbsp;1.1060x** | **0.106753<br>&#x1F7E2;&nbsp;2.2255x** |
| resnet-stem bfloat16 | 0.048032 | **0.066993<br>&#x1F7E2;&nbsp;1.3948x** | **0.053569<br>&#x1F7E2;&nbsp;1.1153x** | **0.106848<br>&#x1F7E2;&nbsp;2.2245x** |
| resnet-stem float32 | 0.056273 | **0.065072<br>&#x1F7E2;&nbsp;1.1564x** | 0.056240<br>&#x1F534;&nbsp;0.9994x | **0.104097<br>&#x1F7E2;&nbsp;1.8499x** |
| vgg-block float16 | 0.009216 | **0.010048<br>&#x1F7E2;&nbsp;1.0903x** | **0.009600<br>&#x1F7E2;&nbsp;1.0417x** | **0.021665<br>&#x1F7E2;&nbsp;2.3508x** |
| vgg-block bfloat16 | 0.009216 | **0.009856<br>&#x1F7E2;&nbsp;1.0694x** | **0.009504<br>&#x1F7E2;&nbsp;1.0312x** | **0.021760<br>&#x1F7E2;&nbsp;2.3611x** |
| vgg-block float32 | 0.012896 | **0.013296<br>&#x1F7E2;&nbsp;1.0310x** | 0.012656<br>&#x1F534;&nbsp;0.9814x | **0.022016<br>&#x1F7E2;&nbsp;1.7072x** |
| alexnet-ceil float16 | 0.010081 | **0.012704<br>&#x1F7E2;&nbsp;1.2602x** | **0.014080<br>&#x1F7E2;&nbsp;1.3967x** | **0.026848<br>&#x1F7E2;&nbsp;2.6632x** |
| alexnet-ceil bfloat16 | 0.010016 | **0.012640<br>&#x1F7E2;&nbsp;1.2620x** | **0.013920<br>&#x1F7E2;&nbsp;1.3898x** | **0.026912<br>&#x1F7E2;&nbsp;2.6869x** |
| alexnet-ceil float32 | 0.013632 | **0.014304<br>&#x1F7E2;&nbsp;1.0493x** | **0.015952<br>&#x1F7E2;&nbsp;1.1702x** | **0.026768<br>&#x1F7E2;&nbsp;1.9636x** |

### MaxPool3dFwdOp and MaxPool3dIndicesFwdOp

| Workload | Candidate (ms) | TileOPs incumbent (ms)<br>/ candidate | torch-compile (ms)<br>/ candidate | torch-ref (ms)<br>/ candidate |
| --- | ---: | ---: | ---: | ---: |
| c3d-pool1 float16 | 0.072896 | **0.072912<br>&#x1F7E2;&nbsp;1.0002x** | **0.074528<br>&#x1F7E2;&nbsp;1.0224x** | **0.405057<br>&#x1F7E2;&nbsp;5.5566x** |
| c3d-pool2 float16 | 0.023168 | 0.023104<br>&#x1F534;&nbsp;0.9972x | **0.026912<br>&#x1F7E2;&nbsp;1.1616x** | **0.076800<br>&#x1F7E2;&nbsp;3.3149x** |
| medicalnet-stem bfloat16 | 0.077217 | **0.085840<br>&#x1F7E2;&nbsp;1.1117x** | **0.702914<br>&#x1F7E2;&nbsp;9.1031x** | **0.259137<br>&#x1F7E2;&nbsp;3.3560x** |
| c3d-pool1 (indices) float16 | 0.118208 | **0.120192<br>&#x1F7E2;&nbsp;1.0168x** | **0.121280<br>&#x1F7E2;&nbsp;1.0260x** | **0.404833<br>&#x1F7E2;&nbsp;3.4248x** |
| c3d-pool2 (indices) float16 | 0.026432 | **0.026944<br>&#x1F7E2;&nbsp;1.0194x** | **0.027969<br>&#x1F7E2;&nbsp;1.0581x** | **0.076896<br>&#x1F7E2;&nbsp;2.9092x** |
| medicalnet-stem (indices) bfloat16 | 0.101088 | **0.188656<br>&#x1F7E2;&nbsp;1.8663x** | **1.543654<br>&#x1F7E2;&nbsp;15.2704x** | **0.259297<br>&#x1F7E2;&nbsp;2.5651x** |

### AvgPool3dFwdOp

| Workload | Candidate (ms) | TileOPs incumbent (ms)<br>/ candidate | torch-compile (ms)<br>/ candidate | torch-ref (ms)<br>/ candidate |
| --- | ---: | ---: | ---: | ---: |
| video-2x2x2 float16 | 0.003008 | **0.003360<br>&#x1F7E2;&nbsp;1.1170x** | **0.004128<br>&#x1F7E2;&nbsp;1.3723x** | **0.008096<br>&#x1F7E2;&nbsp;2.6915x** |
| ceil-video float16 | 0.002943 | **0.003712<br>&#x1F7E2;&nbsp;1.2613x** | **0.003232<br>&#x1F7E2;&nbsp;1.0982x** | **0.009024<br>&#x1F7E2;&nbsp;3.0663x** |
| divisor bfloat16 | 0.001824 | **0.002016<br>&#x1F7E2;&nbsp;1.1053x** | 0.001712<br>&#x1F534;&nbsp;0.9386x | **0.004160<br>&#x1F7E2;&nbsp;2.2807x** |

## Result And Limitations

**SOTA on 16 of 18 2d rows and every 3d row**

- Sixteen of the eighteen 2d rows and all nine 3d and average-pool rows are ahead of every
  comparator, and no row is behind the incumbent by more than 0.5%. `resnet-stem`, the row
  this family entered at 0.72x of torch-compile, now leads it by 1.14x to 1.16x on values
  and 1.00x to 1.12x on indices. What is left behind is `vgg-block` in float32, 0.96x on
  values and 0.98x on indices.
- Where the time went, measured on `resnet-stem` float16 with the flat schedule: the input
  stream alone is 14.58us, nine taps with a plain `>` are 28.83us, and the NaN-propagating
  update takes it to 34.67us. Loads and the update are the whole cost, so both were cut.
  The same measurement on `vgg-block` says the opposite — four taps cost 7.13us against a
  7.17us NaN-propagating version, so there the window is free and only the memory matters,
  which is why the two shapes took different changes.
- The value tile cuts loads. Neighbouring outputs share input columns, so reading the band
  once and separating the window drops loads per output from `kernel_h * kernel_w` to
  `band_h * band_w / (tile_h * tile_w)` — 9 to 5.6 at a 4 by 2 tile. A tile is only offered
  where the band it reads is the band its outputs use, so an axis whose stride outruns its
  kernel, or whose dilation outruns its stride, keeps a tile of one.
- The vector band read cuts transactions. It needs a band that is a whole number of 16-byte
  runs on a 16-byte boundary, which a 3 by 3 window at stride 2 never has — its band is
  `2 * tile_w + 1` columns, always odd. `vgg-block` is the workload it serves, and it takes
  float16 there from 6.77us to 6.11us.
- The with-indices kernel drops three of its five accumulators. PyTorch decides with
  `val > maxval || isnan(val)` and seeds the position with the window's first tap inside
  the input; that one predicate already gives the first maximum on a tie, the last NaN in
  the window, and the first valid position for a window holding nothing but -inf, which is
  what `has_nan`, `nan_idx` and `first_valid` were reconstructing.
- `vgg-block` float32 was chased and did not move. A chunked 16-byte band at a wider tile,
  a vectorised store, and tile heights up to 4 were each measured; the best of them is the
  10.368us in the table. The shape reads 32.11 MB at 3.10 TB/s against an elementwise
  ceiling of 3.64 TB/s measured the same way, and one load instruction already covers 512
  contiguous bytes with every byte used, so the remaining 4% is not a load-shape question
  this PR can answer.
- Rejected on measurement, `resnet-stem` float16 with everything else held fixed: a
  shared-memory staged patch 65.44us; an interior fast path over the tiled body 35.46us;
  `tl.max_nan` 36.99us, and it is wrong for float32, which it does not serve; a 4 by 4 tile
  36.74us against 4 by 2's 30.69us; a per-thread interior claim 48.19us; a per-CTA one
  43.84us. Hoisting the clamp to a per-thread window base is not a trade-off but a bug:
  clamping the first column slides every tap of the window, so the taps that survive read
  the wrong columns.
- The guard is not the same shape at both ranks, and the two were measured against each
  other. A nine-tap 2d window keeps the warp together and lets the column test ride the
  update: 34.56us against 41.25us for a branch. A twenty-seven-tap 3d window skips the tap
  under one combined predicate: 82.23us against 95.20us.
- The with-indices kernels carry the running position in `int32`, as they did before this
  PR. That overflows once a plane exceeds `2**31` elements — `H_in * W_in` at 2d,
  `D_in * H_in * W_in` at 3d — which no manifest workload approaches and which this PR does
  not change.
- 263 `tests/ops/test_pool.py` cases pass. The file gains one node, 262 -> 263: the 2d
  parameter set had padding on every case, so the branch taken when a window provably stays
  inside the input was never reached, and the added case is the first shape that reaches it.

### CODE REVIEW

File: src/tileops/kernels/pool/max_pool2d.py [hunk 1]
```diff
@@ -32,52 +42,222 @@ def _max_pool2d_kernel(
...
+                                    # Neighbouring threads hold neighbouring columns,
+                                    # so a branch here splits the warp at the row
+                                    # edges. The column test rides the update instead.
+                                    v = T.cast(x[row, ih, safe_w(iw)], accum_dtype)
+                                    live = cols_inside or ((iw >= 0) and (iw < w_in))
+                                    # NaN enters `run` and never leaves: a later
+                                    # value fails `v > NaN`, which is how PyTorch
+                                    # propagates it.
+                                    run = T.if_then_else(live and (T.isnan(v) or (v > run)), v, run)
+                        out[row, oh, ow] = T.cast(run, dtype)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply src/tileops/kernels/pool/max_pool2d.py]: Checked, and the premise is wrong in the other direction: the scalar is fine and the array was unnecessary, so 49195116 removes it and the 3d kernels now match the 1d and 2d form.

The 3d value scan is three plain nested `T.serial` loops with one predicate inside them — the interleaving the comment described was in an earlier draft that this PR does not ship. The miscomputation that prompted the array was the build-time bool collapsing the `or` in the with-indices update, which `seed = False if window_inside else first_valid` already fixes; swapping the accumulator did not fix it at the time and I left the array in defensively without re-testing the claim.

Evidence for the swap: 263 `tests/ops/test_pool.py` cases pass, and the probe that originally caught the bug reports 0 mismatched values and 0 mismatched positions on the `pad=0, k=2, s=2` shape. Timing is unchanged — TileLang keeps a one-element local in a register — `medicalnet-stem` 77.17us against 77.25us on values and 175.33us against 175.39us on indices, both inside run-to-run spread.
